### ISSUE TITLE
[Potential Bug] Blocksparse fwd_block call requires q, k, v separated, but API call passes fused qkv matrix.

### ISSUE THREAD

**Issue by codedecde** [HazyResearch/flash-attention#66]
Hi 
I was trying out the blocksparse flash attention kernel, and had some questions about it.

### Potentially Incorrect API for fwd_block

[This line](https://github.com/HazyResearch/flash-attention/blob/9b0bc978729d5470aaf5667b8f65df6fa2b3a007/flash_attn/flash_blocksparse_attn_interface.py#L45) passes the fused qkv matrix to the fwd_block function

```[python]
context, softmax_lse, *rest = flash_attn_cuda.fwd_block(qkv, cu_seqlens, blockmask, dropout_p,
                                                             max_s, softmax_scale, causal,
                                                             return_softmax, None)
```

However, looking at the code [here](https://github.com/HazyResearch/flash-attention/blob/9b0bc978729d5470aaf5667b8f65df6fa2b3a007/csrc/flash_attn/fmha_api.cpp#L492), the `fwd_block` expects the following signature:
```[cpp]
mha_fwd_block(const at::Tensor &q,         // total_q x num_heads x head_size, total := \sum_{i=0}^{b} s_i
              const at::Tensor &k,         // total_k x num_heads x head_size, total_k := \sum_{i=0}^{b} s_i
              const at::Tensor &v,         // total_k x num_heads x head_size, total_k := \sum_{i=0}^{b} s_i
              const at::Tensor &cu_seqlens_q,  // b+1
              const at::Tensor &cu_seqlens_k,  // b+1
              const at::Tensor &blockmask,   // (seqlen / 256, seqlen / 16)
              const int max_seqlen_q_,
              const int max_seqlen_k_,
              const float p_dropout,
              const float softmax_scale,
              const bool is_causal,
              const bool return_softmax,
              c10::optional<at::Generator> gen_)
```
So, (to the best of my understanding), shouldn't the api call be with the q, k and v matrices split (along with the cu_seqlens_k part etc.) ?

### Support for Bfloat16
Right now the type-checks enforce fp16 checks ([this line](https://github.com/HazyResearch/flash-attention/blob/9b0bc978729d5470aaf5667b8f65df6fa2b3a007/csrc/flash_attn/fmha_api.cpp#L512)). Changing the checks to be similar to `mha_fwd` ([here](https://github.com/HazyResearch/flash-attention/blob/9b0bc978729d5470aaf5667b8f65df6fa2b3a007/csrc/flash_attn/fmha_api.cpp#L206)) seems at least have the forward function go through, but is there anything else that needs to be changed ?

### Support for Non Causal Layouts
[This line](https://github.com/HazyResearch/flash-attention/blob/9b0bc978729d5470aaf5667b8f65df6fa2b3a007/flash_attn/flash_blocksparse_attn_interface.py#L21) indicates that causal block layouts are not supported. Is there a specific reason for that ?

**Comment by tridao** [comment]
Thanks! We haven't spent as much time on the block sparse code as the standard attention, so there could be mismatches as the code evolved.
- Yes, I think the API should call q, k, v separately
- To get bf16 working with block sparse, one would also need to change the [kernel](https://github.com/HazyResearch/flash-attention/blob/main/csrc/flash_attn/src/fmha_block_fprop_kernel_1xN.h) to template on a type. Right now it hard-codes the type to be __half (i.e. fp16). For example the standard attention [kernel](https://github.com/HazyResearch/flash-attention/blob/main/csrc/flash_attn/src/fmha_fprop_kernel_1xN.h)  templates on the type `elem_type` and then dispatches with either `elem_type = __half` or `elem_type = __nv_bfloat16`.
- No specific reason that block sparse doesn't support causal, it's just we haven't implemented / tested that code path.

**Comment by codedecde** [comment]
Thank you for the quick response ! 

I made the changes, and at least the forward and backward execute okay. I was trying to run some sanity checks for the blocksparse, but am running into issues. Specifically, I'm running the following test (with modifications to the API).

```[python]
import math
from typing import Optional
import pathlib
import torch
from flash_attn.flash_blocksparse_attn_interface import convert_blockmask
from flash_attn.flash_attn_interface import flash_attn_func
import flash_attn_cuda
import torch.nn.functional as F

def collapse_first_n_dims(x, n):
    new_shape = (-1,) + x.shape[n:]
    return x.view(new_shape)


############# Packed Version of the Code ##############

def _flash_blocksparse_attn_forward(
    qkv: torch.Tensor,
    cu_seqlens: torch.IntTensor,
    blockmask: torch.IntTensor,
    dropout_p: float,
    max_s: int,
    softmax_scale: float,
    causal: bool,
    return_softmax: bool,
):
    # the fwd block takes the q, k and v as separate tensors.
    # so we pass them as such
    q, k, v = qkv[:, 0], qkv[:, 1], qkv[:, 2]
    cu_seqlens_q = cu_seqlens
    cu_seqlens_k = cu_seqlens
    max_seqlen_q, max_seqlen_k = max_s, max_s
    context, softmax_lse, *rest = flash_attn_cuda.fwd_block(
        q,  # (bs * seq_len, num_heads, head_size)
        k,  # (bs * seq_len, num_heads, head_size)
        v,  # (bs * seq_len, num_heads, head_size)
        cu_seqlens_q,  # (bs + 1,)
        cu_seqlens_k,  # (bs + 1,)
        blockmask,  # (seqlen // 256, seqlen // 16), 16 is the row block size, 256 is the column block size
        max_seqlen_q,
        max_seqlen_k,
        dropout_p,
        softmax_scale,
        causal,
        return_softmax,
        None,
    )
    # if context.isnan().any() or softmax_lse.isnan().any():
    #     breakpoint()
    S_dmask = rest[0] if return_softmax else None
    return context, softmax_lse, S_dmask


def _flash_blocksparse_attn_backward(
    dout: torch.Tensor,
    qkv: torch.Tensor,
    out: torch.Tensor,
    dq: torch.Tensor,
    dk: torch.Tensor,
    dv: torch.Tensor,
    softmax_lse: torch.Tensor,
    cu_seqlens: torch.IntTensor,
    blockmask: torch.IntTensor,
    dropout_p: float,
    max_s: int,
    softmax_scale: float,
    causal: bool
) -> torch.Tensor:
    q, k, v = qkv[:, 0], qkv[:, 1], qkv[:, 2]
    cu_seqlens_q, cu_seqlens_k = cu_seqlens, cu_seqlens
    max_seqlen_q, max_seqlen_k = max_s, max_s
    dout = dout.contiguous()
    softmax_d = flash_attn_cuda.bwd_block(
        dout,  # (bs * seq_len, num_heads, head_size)
        q,  # (bs * seq_len, num_heads, head_size)
        k,  # (bs * seq_len, num_heads, head_size)
        v,  # (bs * seq_len, num_heads, head_size)
        out,
        softmax_lse,
        dq,
        dk,
        dv,
        cu_seqlens_q,
        cu_seqlens_k,
        blockmask,
        max_seqlen_q,
        max_seqlen_k,
        dropout_p,
        softmax_scale,
        causal,
        None
    )
    # if dqkv.isnan().any() or softmax_d.isnan().any():
    #     breakpoint()
    return dq, dk, dv, softmax_d

class FlashBlocksparseAttnFun(torch.autograd.Function):

    @staticmethod
    def forward(ctx, qkv, cu_seqlens, blockmask, dropout_p, max_s, softmax_scale, causal):
        # Save rng_state because the backward pass will regenerate the dropout mask
        rng_state = torch.cuda.get_rng_state() if dropout_p > 0 else None
        if softmax_scale is None:
            softmax_scale = qkv.shape[-1] ** (-0.5)
        context, softmax_lse, S_dmask = _flash_blocksparse_attn_forward(
            qkv, cu_seqlens, blockmask, dropout_p, max_s, softmax_scale, causal=causal,
            return_softmax=False
        )
        ctx.save_for_backward(qkv, context, S_dmask, softmax_lse, cu_seqlens, blockmask, rng_state)
        ctx.dropout_p = dropout_p
        ctx.max_s = max_s
        ctx.softmax_scale = softmax_scale
        ctx.causal = causal
        return context

    @staticmethod
    def backward(ctx, dout):
        qkv, context, S_dmask, softmax_lse, cu_seqlens, blockmask, rng_state = ctx.saved_tensors
        if rng_state is not None:
            cur_rng_state = torch.cuda.get_rng_state()
            torch.cuda.set_rng_state(rng_state)
        # S_dmask is None, temporarily use another tensor just to get it running
        dqkv = torch.empty_like(qkv)
        _flash_blocksparse_attn_backward(
            dout=dout,
            qkv=qkv,
            out=context,
            dq=dqkv[:, 0],
            dk=dqkv[:, 1],
            dv=dqkv[:, 2],
            softmax_lse=softmax_lse,
            cu_seqlens=cu_seqlens,
            blockmask=blockmask,
            dropout_p=ctx.dropout_p,
            max_s=ctx.max_s,
            softmax_scale=ctx.softmax_scale,
            causal=ctx.causal,
        )
        if rng_state is not None:
            torch.cuda.set_rng_state(cur_rng_state)
        return dqkv, None, None, None, None, None, None, None


# We duplicate code to return both the output and the softmax for testing
# Returning both makes backward a bit slower, so we want to keep using the other version for speed.
class FlashBlocksparseAttnFunWithS(torch.autograd.Function):

    @staticmethod
    def forward(ctx, qkv, cu_seqlens, blockmask, dropout_p, max_s, softmax_scale, causal):
        # Save rng_state because the backward pass is gonna regenerate the dropout mask
        rng_state = torch.cuda.get_rng_state() if dropout_p > 0 else None
        if softmax_scale is None:
            softmax_scale = qkv.shape[-1] ** (-0.5)
        context, softmax_lse, S_dmask = _flash_blocksparse_attn_forward(
            qkv, cu_seqlens, blockmask, dropout_p, max_s, softmax_scale, causal=causal,
            return_softmax=True
        )
        ctx.save_for_backward(qkv, context, S_dmask, softmax_lse, cu_seqlens, blockmask, rng_state)
        ctx.dropout_p = dropout_p
        ctx.max_s = max_s
        ctx.softmax_scale = softmax_scale
        ctx.causal = causal
        return context, S_dmask, softmax_lse

    @staticmethod
    def backward(ctx, dout, *args):
        qkv, context, S_dmask, softmax_lse, cu_seqlens, blockmask, rng_state = ctx.saved_tensors
        if rng_state is not None:
            cur_rng_state = torch.cuda.get_rng_state()
            torch.cuda.set_rng_state(rng_state)
        dqkv = _flash_blocksparse_attn_backward(
            dout, qkv, context, S_dmask, softmax_lse, cu_seqlens, blockmask, ctx.dropout_p,
            ctx.max_s, ctx.softmax_scale, ctx.causal
        )
        if rng_state is not None:
            torch.cuda.set_rng_state(cur_rng_state)
        return dqkv, None, None, None, None, None, None
    
def flash_blocksparse_attn_func(
    qkv: torch.Tensor,
    cu_seqlens: torch.IntTensor,
    blockmask: torch.IntTensor,
    dropout_p: float,
    max_s: int,
    softmax_scale: Optional[float] = None,
    causal: bool = False,
    return_attn_probs: bool = False,
    convert_mask: bool = True,
) -> torch.Tensor:
    func = FlashBlocksparseAttnFun if not return_attn_probs else FlashBlocksparseAttnFunWithS
    if convert_mask:
        blockmask = convert_blockmask(blockmask, causal=causal)
    return func.apply(
        qkv,
        cu_seqlens,
        blockmask,
        dropout_p,
        max_s,
        softmax_scale,
        causal,
    )

######### Unit tests start ########

def test_blocksparse_attn_fwd_bwd():
    # This passes okay
    root = pathlib.Path(__file__).parent
    fixture_root = root / 'fixtures'
    fixture_path = fixture_root / "flash_attn_input.pt"
    assert fixture_path.exists()
    inp = torch.load(fixture_path)
    device = torch.device("cuda:0")
    qkv = inp["qkv"].to(device)
    qkv.requires_grad = True
    cu_seqlens = inp["cu_seqlens"].to(device)
    blockmask = inp["blockmask"].to(device)
    dropout_p = inp["dropout_p"]
    max_s = inp["max_s"]
    softmax_scale = inp["softmax_scale"]
    causal = inp["causal"]
    convert_mask = inp["convert_mask"]
    context_layer = flash_blocksparse_attn_func(
        qkv=collapse_first_n_dims(qkv, 2),
        cu_seqlens=cu_seqlens,
        blockmask=blockmask,
        dropout_p=dropout_p,
        max_s=max_s,
        softmax_scale=softmax_scale,
        causal=causal,
        return_attn_probs=False,
        convert_mask=convert_mask,
    )
    item = context_layer.sum()
    item.backward()
    print("Forward and backward executed successfully")


def test_blocksparse_dense_match():
    # This test case fails right now :/
    bs = 1
    hdim = 16
    sq = 256
    n_heads = 1
    device = torch.device("cuda:0")
    def repeat_along_dim(arr, n_times, dim):
        return arr.repeat(*(1 if i != dim else n_times for i in range(arr.ndim)))

    def create_query_key_value():
        key = torch.zeros(bs, sq, hdim)
        for ix in range(min(sq, hdim)):
            key[:, ix, ix] = 1
        query = torch.zeros(bs, sq, hdim)
        query[:, 0, 1] = 1.0
        query[:, 1, 0] = 1.0
        for idx in range(2, min(hdim, sq)):
            query[:, idx, idx] = 1.0
        value = torch.arange(bs * sq * hdim, dtype=torch.bfloat16).reshape((bs, hdim, sq)).transpose(1, 2)
        
        query = repeat_along_dim(query.unsqueeze(2), n_heads, 2)
        key = repeat_along_dim(key.unsqueeze(2), n_heads, 2)
        value = repeat_along_dim(value.unsqueeze(2), n_heads, 2)
        qkv = torch.cat([query.unsqueeze(2), key.unsqueeze(2), value.unsqueeze(2)], dim=2)
        return qkv.to(device=device, dtype=torch.bfloat16)
    # shape: (bs, sq, 3, nh, hdim)
    qkv = create_query_key_value()
    # shape: (bs, sq, nh, hdim)
    query = qkv[:, :, 0]
    # shape: (bs, sq, nh, hdim)
    key = qkv[:, :, 1]
    # shape: (bs, sq, nh, hdim)
    value = qkv[:, :, 2]

    assert query.shape[2] == key.shape[2] == value.shape[2] == n_heads
    # shape: (bs, nh, sq, sq)
    dense_logits = torch.einsum("bsnh,bqnh -> bnsq", query, key) * 1.0 / math.sqrt(hdim)
    # shape: (bs, nh, sq, sq)
    dense_scores = torch.softmax(dense_logits, dim=-1)
    # shape: (bs, nh, sq, hdim)
    dense_values = torch.einsum("bnst,btnd -> bnsd", dense_scores, value)

    # Using FlashAttention Interface

    bs, sq = qkv.size(0), qkv.size(1)
    cu_seqlens = torch.arange(0, (bs + 1)*sq, sq, device=device, dtype=torch.int32)
    flash_attn_dense_values = flash_attn_func(
        qkv=collapse_first_n_dims(qkv.clone(), 2),
        cu_seqlens=cu_seqlens,
        dropout_p=0.0,
        max_s=sq,
        softmax_scale=None,
        causal=False,
        return_attn_probs=False,
    ).reshape((bs, sq, n_heads, hdim))

    # Using blocksparse interface
    row_block_size = 16
    col_block_size = 256
    
    def create_layout():
        layout = torch.ones((sq // row_block_size, sq // col_block_size), dtype=torch.int32, device=device)
        return layout
    # shape: (sq // row_block_size, sq // col_block_size)
    layout = create_layout()
    blockmask = convert_blockmask(layout, causal=False)
    
    flash_attn_sparse_values = flash_blocksparse_attn_func(
        qkv=collapse_first_n_dims(qkv.clone(), 2),
        cu_seqlens=cu_seqlens,
        blockmask=blockmask,
        dropout_p=0.0,
        max_s=sq,
        softmax_scale=None,
        causal=False,
        return_attn_probs=False,
        convert_mask=False,
    ).reshape((bs, sq, n_heads, hdim))
    # assert torch.allclose(flash_attn_dense_values, flash_attn_sparse_values)
    import pdb; pdb.set_trace()
    


if __name__ == "__main__":
    test_blocksparse_attn_fwd_bwd()
    test_blocksparse_dense_match()
```

In the above, I'm basically trying to compute the attention scores with the dense version, and then with the sparse variant with an all 1's layout. I think the outputs should have been the same between them, but there is a substantial difference between `flash_attn_sparse_values` and `flash_attn_dense_values`

```[bash]
(Pdb) flash_attn_dense_values - flash_attn_sparse_values
tensor([[[[ 35.0000,  -2.0000,   0.0000,  ...,   0.0000,   0.0000, -16.0000]],

         [[ 38.0000,  -2.0000,   0.0000,  ...,   0.0000,   0.0000, -16.0000]],

         [[ 35.0000,  -2.0000,   0.0000,  ...,   0.0000,   0.0000, -16.0000]],

         ...,

         [[ 33.5000,   0.0000,   0.0000,  ...,   0.0000,   0.0000,   0.0000]],

         [[ 33.5000,   0.0000,   0.0000,  ...,   0.0000,   0.0000,   0.0000]],

         [[ 33.5000,   0.0000,   0.0000,  ...,   0.0000,   0.0000,   0.0000]]]],
       device='cuda:0', dtype=torch.bfloat16)
```

I'm not sure what might be going wrong .. Would be grateful for any help ! :)

**Comment by Fword4u** [comment]
I met this problem too.I modified the code to make the sparse block attention work but I get a output tensor filled with zero.I checked the block mask and anything else, still no idea
